scripts/algorithms/fm.dml

```diff
@@ -0,0 +1,80 @@
+#-------------------------------------------------------------
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+# 
+#   http://www.apache.org/licenses/LICENSE-2.0
+# 
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+# Implements the (2-way) factorization machines.
+#
+# ------------------------------------------------------------------------------------------------------
+# NAME      TYPE     DEFAULT  MEANING
+# ------------------------------------------------------------------------------------------------------
+# X         String   ---      Location to read the input matrix X to be factorized
+# y         String   ---      Location to read the input matrix y corresponding to the feature vector
+# k         Int      ---      hyperparameter that defines the dimensionality of factorization
+# -------------------------------------------------------------------------------------------------------
+#
+# OUTPUT: Updated Parameters w0, W, V, in matrix form. 
+# 
+
+fileX = $X; # nxm : n rows and m columns
+fileY = $y; # nx1
+
+X = read (fileX);
+y = read (fileY);
+
+w0 = matrix(0, rows=1, cols=1); # 1x1
+W  = matrix(0, rows=1, cols=n); # 1xn
+V  = rand(rows=n, cols=k, min=0.0, max=1.0); # nxk
+
+# model equation 
+# Y = w0 + sum(W * X) + 0.5 * sum( ( X %*% t(V) ) - ( (X) ^ 2 ) %*% t( V ^ 2));
```

Comment:
In the paper, the equations are for a single example, so for all examples, we will want `Y` (or we could name it `y` since it will be a column vector) to be of shape `(N, 1)` (or `(n, 1)` if we choose the cleaner notation with `n` examples.  Therefore, this should reduce down to

```
y = w0 + X %*% W + 0.5 * rowSums( (X %*% V)^2 - (X^2 %*% V^2) )
```

where

* `y` is of shape `(n, 1)`, 
* `w_0` -> `(1,)`, 
* `X %*% W` -> `(n, 1)`, 
* `(X %*% V)^2` -> `(n, k)`, 
* `X^2  %*% V^2` -> `(n, k)`,
* and `rowSums( (X %*% V)^2 - X^2 %*% V^2 )` -> `(n, 1)`,

using the updated notation with `n` examples and `d` features.